Testing with Datadog!

astuyve opened the pull request:
Hey folks! This PR adds instructions on how to test a Serverless application with Datadog APM, Synthetics, and Monitors!

Feel free to make any suggestions.  This PR is ready for review.

Thanks!

brnkrygs commented:
Thanks for the PR @astuyve! We have a couple of folks tasked to review as soon as possible. Appreciate the effort and partnership.

astuyve commented:
Hey @brnkrygs - one problem I just realized is that the `.gitignore` file is ignoring the `lib` directory, which causes my CDK stack not to be committed. I can force-push it, but perhaps we could re-visit [this configuration](https://github.com/aws-samples/serverless-test-samples/blob/main/.gitignore#L128)?

dancfox reviewed (commented):
cdk deploy gives me the following error:

Error: Cannot find module './apigw-lambda-sqs-sns-datadog.ts'

astuyve commented:
@dancfox - thanks for the note!

Could you try pulling this with an updated git ignore which removes `lib` from the ignore path? Right now the entire project's `.gitignore` will filter out any`lib` directory, I referenced that [here](https://github.com/aws-samples/serverless-test-samples/pull/203#issuecomment-1533516495). I didn't want to remove that setting as I'm not sure why it was added. So I suspect either you may have an older version of this branch, or `git pull` is ignoring the file.

I can push a new commit which drops this check if that helps.

Thanks!

dancfox commented:
https://github.com/aws-samples/serverless-test-samples/commit/8554fc12c71f209d2060883f228c5aae07d68abc

astuyve commented:
@dancfox Aha! also files in `/bin` directories are ignored

brnkrygs commented:
> 

Hi @astuyve - we ran into this in a separate PR as well... anything with CDK will hit it.

To get around it, you can directly add the file with `git add ...path..to...file...to...add` even though `.gitignore` is ... ignoring it, but looks like @dancfox beat me to it with a `.gitignore` update.

Given the mix of frameworks in this repo we may have to go back and forth a bit to tune this, or include some explicit instructions.

Note that you'll probably have to explicitly add a file under `/bin` as well.

Check out the thread here for our earlier troubleshooting:
https://github.com/aws-samples/serverless-test-samples/pull/204/files/d4fefcd15a5a6a783980711a9d024df3a8e07edd#diff-19191bb47680afb01355baf82f2cf99a2a74c3537a31da851b7455b09dc99a56

astuyve commented:
Hi @brnkrygs - thanks for the note! I did stumble into that and pushed the file manually about 30s ago in this commit: https://github.com/aws-samples/serverless-test-samples/pull/203/commits/ec1d80d8f53ce26366143e1c5db486b54d9c311d

I pulled this into a new directory and was able to build the stack, so we should be set there. Please let me know if there are further issues, thank you!

brnkrygs commented on typescript-test-samples/apigw-lambda-sqs-sns-datadog/README.md:
```diff
@@ -0,0 +1,76 @@
+6. Deploy this example to AWS! Run `DD_API_KEY=<Your API Key> cdk deploy`
+
+## Testing
+1. Manually curl the endpoint to verify things work:
```
Getting Internal server error here.

CloudWatch logs for the Lambda function show this:

```2023-05-08T18:11:23.610Z	undefined	ERROR	Uncaught Exception 	{
    "errorType": "Runtime.ImportModuleError",
    "errorMessage": "Error: Cannot find module 'publisher'\nRequire stack:\n- /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js\n- /opt/nodejs/node_modules/datadog-lambda-js/runtime/index.js",
    "stack": [
        "Runtime.ImportModuleError: Error: Cannot find module 'publisher'",
        "Require stack:",
        "- /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js",
        "- /opt/nodejs/node_modules/datadog-lambda-js/runtime/index.js",
        "    at ImportModuleError.ExtendedError [as constructor] (/opt/nodejs/node_modules/datadog-lambda-js/runtime/errors.js:113:28)",
        "    at new ImportModuleError (/opt/nodejs/node_modules/datadog-lambda-js/runtime/errors.js:123:42)",
        "    at /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:273:31",
        "    at step (/opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:43:23)",
        "    at Object.throw (/opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:24:53)",
        "    at rejected (/opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js:16:65)"
    ]
}
```

dancfox replied:
When I run the test in APIGW I get this error:

```
Mon May 08 18:21:22 UTC 2023 : Lambda execution failed with status 200 due to customer function error: Error: Cannot find module 'axios'
Require stack:
- /var/task/publisher.js
- /opt/nodejs/node_modules/datadog-lambda-js/runtime/user-function.js
- /opt/nodejs/node_modules/datadog-lambda-js/runtime/index.js. Lambda request id: 99657aac-2f6e-4c27-8852-7b36183703fa
```

astuyve replied:
Ah looks like `npm install` and `npm run build` won't install the dependencies in the `resources` directory. Should I lift axios to be a top level dependency then? I thought that CDK's `assetFrom` method required all deps to be in the same directory. Maybe there's a canonical fix for this here?

This should work if you run `npm install` manually in the `resources` directory.

brnkrygs replied:
Running `npm install` from the `resources` directory did the trick for me.

As far as a "canonical fix"....

This pattern uses CDK, we could potentially use a similar config:
https://github.com/aws-samples/serverless-test-samples/tree/main/typescript-test-samples/lambda-handler-dynamodb

IIRC its based off of `cdk init` and doesn't use a separate `resources` package - just the top-level one. Personally, I've used way more SAM than CDK, so I'll ask internally for another pair of eyes.

astuyve replied:
Okay, I can also update the readme or even use `fetch` instead of axios. I'm also not extremely familiar with the CDK and mostly use the Serverless Framework, so it's a bit unclear in this context how to differentiate from runtime dependencies from build time dependencies - however if putting all of the application code at the root level works, we could do that.

Let me know, thanks!

brnkrygs replied:
CDK best practices include [keeping infra and runtime code in the same package](https://docs.aws.amazon.com/cdk/v2/guide/best-practices.html#best-practices-constructs).

CDK init puts runtime code in a subdirectory, but uses shared root-level dependencies (single `package.json` in the root).

To avoid huge functions, the CDK `aws-cdk-lib/aws-lambda-nodejs/NodejsFunction` construct uses `esbuild` (or Docker) to tree-shake and bundle individual functions.

Suggestions here (will comment on the specific files as well for clarity):

1. Remove `/resources/package.json`
2. In `/lib/apigw-lambda-sqs-sns-datadog-stack.ts`, import `NodejsFunction` and use it instead of `lambda.Function` as seen [in this example](https://github.com/aws-samples/serverless-test-samples/blob/076d8652936124cabf1e1a98a9c9b47f215d3dff/typescript-test-samples/lambda-handler-dynamodb/lib/lambda-handler-dynamodb-stack.ts#L8)
    1. Set `entry` property instead of `code` property
    2. Adjust `handler` property to drop module name - e.g. `handler: "main"` instead of `handler: "publisher.main"`
5. Confirm `./package.json` includes axios (looks like it does at the moment)
6. Install a devDependency on `esbuild` to [avoid requiring Docker](https://docs.aws.amazon.com/cdk/api/v1/docs/aws-lambda-nodejs-readme.html#local-bundling)

These tweaks keep the bulk of the app structure as-is, which is aligned to CDK recommended structure, and also allow for the nice and simple README instructions you have already.

brnkrygs commented on typescript-test-samples/apigw-lambda-sqs-sns-datadog/resources/package.json:
Per thread on README.md, drop this.

brnkrygs commented on typescript-test-samples/apigw-lambda-sqs-sns-datadog/package.json:
```diff
@@ -0,0 +1,33 @@
+    "jest": "^29.5.0",
+    "ts-jest": "^29.0.5",
+    "ts-node": "^10.9.1",
+    "typescript": "~4.9.5"
```
Add `esbuild` as `devDependency` to avoid Docker requirement.

https://docs.aws.amazon.com/cdk/api/v1/docs/aws-lambda-nodejs-readme.html#local-bundling

brnkrygs commented on typescript-test-samples/apigw-lambda-sqs-sns-datadog/lib/apigw-lambda-sqs-sns-datadog-stack.ts:
```diff
@@ -0,0 +1,63 @@
+import * as snsSubscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
+import * as lambdaEventSources from 'aws-cdk-lib/aws-lambda-event-sources';
+import { Datadog } from "datadog-cdk-constructs-v2";
+
```
Add import for `NodejsFunction` e.g.:
`import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';`

brnkrygs commented on typescript-test-samples/apigw-lambda-sqs-sns-datadog/lib/apigw-lambda-sqs-sns-datadog-stack.ts:
```diff
@@ -0,0 +1,63 @@
+      visibilityTimeout: cdk.Duration.seconds(300)
+    });
+
+    const publisherHandler = new lambda.Function(this, "PublisherHandler", {
```
Use `NodejsFunction` to leverage `esbuild` bundler.

brnkrygs commented on typescript-test-samples/apigw-lambda-sqs-sns-datadog/lib/apigw-lambda-sqs-sns-datadog-stack.ts:
```diff
@@ -0,0 +1,63 @@
+      },
+    });
+
+    const workerHandler = new lambda.Function(this, "WorkerHandler", {
```
Use `NodejsFunction` to leverage `esbuild` bundler.

brnkrygs reviewed (changes requested):
Hi @astuyve - have some comments/suggestions for you on this PR around the CDK build/deploy issue. Happy to discuss further if needed!

Thanks again for your effort.

astuyve commented:
Hey thanks, hopefully I'll have an update here in the next couple days!

astuyve commented:
Hy @brnkrygs - sorry for the delay. I believe I've addressed the comments, please feel free to add more if there's something I missed!

Thanks again!

dancfox reviewed (approved):
This looks great. Thank you!
